commcare-cloud-bootstrap/commcare_cloud_bootstrap.py

```diff
@@ -38,73 +49,112 @@ class AwsConfig(jsonobject.JsonObject):
-    instance_ids = ask_aws_for_instances(env, spec.aws_config, len(all_hosts))
+    inventory = bootstrap_inventory(spec, env)
+    instance_ids = ask_aws_for_instances(env, spec.aws_config, len(inventory.all_hosts))
 
     while True:
         instance_ip_addresses = poll_for_aws_state(env, instance_ids)
         if instance_ip_addresses:
             break
 
-    ip_addresses_by_host = dict(zip(all_hosts, instance_ip_addresses.values()))
+    hosts_by_name = {}
+
+    for host, (public_ip, private_ip) in zip(inventory.all_hosts, instance_ip_addresses.values()):
+        host.public_ip = public_ip
+        host.private_ip = private_ip
+        hosts_by_name[host.name] = host
 
-    for group_name, group in inventory.all.children.items():
-        for host_name, host_vars in group.hosts.items():
-            host_vars['ansible_host'] = str(ip_addresses_by_host[host_name])
+    for i, host_name in enumerate(inventory.all_groups['kafka'].host_names):
+        hosts_by_name[host_name].vars['kafka_broker_id'] = i
+        hosts_by_name[host_name].vars['swap_size'] = '2G'
+
+    for host_name in inventory.all_groups['elasticsearch'].host_names:
+        hosts_by_name[host_name].vars['elasticsearch_node_name'] = host_name
 
     save_inventory(env, inventory)
     copy_default_vars(env, spec.aws_config)
 
 
+def alphanumeric_sort_key(key):
+    """
+    Sort the given iterable in the way that humans expect.
+    Thanks to http://stackoverflow.com/a/2669120/240553
+    """
+    import re
+    convert = lambda text: int(text) if text.isdigit() else text
+    return [convert(c) for c in re.split('([0-9]+)', key)]
+
+
 def bootstrap_inventory(spec, env):
-    inventory = AnsibleInventory()
     incomplete = dict(spec.allocations.items())
 
-    all_hosts = set()
+    inventory = Inventory()
 
     while incomplete:
-        for name, allocation in incomplete.items():
+        for role, allocation in incomplete.items():
             if allocation.from_:
                 if allocation.from_ not in spec.allocations:
                     raise KeyError('You specified an unknown group in the from field of {}: {}'
-                                   .format(name, allocation.from_))
+                                   .format(role, allocation.from_))
                 if allocation.from_ in incomplete:
                     continue
-                inventory.all.children[name] = AnsibleInventoryGroup(hosts={
-                    host: {}
-                    for host in inventory.all.children[allocation.from_].hosts.keys()[:allocation.count]
-                })
+                # This is kind of hacky because it does a string sort
+                # on strings containing integers.
+                # Once we have more than 10 it'll start sorting wrong
+                host_names = sorted(
+                    inventory.all_groups[allocation.from_].host_names,
+                    key=alphanumeric_sort_key,
+                )[:allocation.count]
```

Comment:
I'm having trouble following this part - might come ping you in person.

Reply:
Okay, a comment might be in order here :)